File: 백준/Platinum/1168. 요세푸스 문제 2/요세푸스 문제 2.cc
#include<iostream>
#include<algorithm>
using namespace std;

int N, K;
int a, b;
int seg[(1<<18)];

// 세그먼트트리 초기화
int init(int node, int s, int e) // (루트노드, 시작, 끝)
{
    if (s == e) return seg[node] = 1; // start 와 end 의 위치가 일치하면 1을 넣어준다.
    int mid = (s + e) / 2;
    return seg[node] = init(2 * node, s, mid) + init(2 * node + 1, mid + 1, e);
}

// 세그먼트트리 정보 수정
int update(int node, int s, int e, int del) // (루트노드, 시작, 끝, 제거번호)
{
    seg[node]--;
    if (s == e) return 0;
    else
    {
        int mid = (s + e) / 2;
        if (del <= mid) return update(2 * node, s, mid, del);
        else return update(2 * node + 1, mid + 1, e, del);
    }
}

// 다음 순서에 해당하는 번호 받아오기
int query(int node, int s, int e, int order) // (현재노드, 시작, 끝, 순서)
{
    if (s == e) return s; // start 와 end 의 위치가 일치하면 번호를 반환한다.
    int mid = (s + e) / 2;

    if (order <= seg[2 * node]) return query(2 * node, s, mid, order);
    else return query(2 * node + 1, mid + 1, e, order - seg[2 * node]);

}


int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie();

    cin >> N >> K;

    init(1, 1, N); // (루트노드, 시작, 끝)

    int index = 1;

    cout << "<";

    for (int i = 0; i < N; i++)
    {
        // 다음 순서 구하기
        int size = N - i; // 사람 수
        index += K - 1;

        if (index % size == 0) index = size;
        else if (index > size) index %= size;
            

        // 다음 순서에 해당하는 번호 받아오기
        int num = query(1, 1, N, index);

        // 해당 번호 제거 하기
        update(1, 1, N, num);
          
        if (i == N - 1) cout << num;
        else cout << num << ", ";
    }
    cout << ">";
}

File: 백준/Platinum/1168. 요세푸스 문제 2/요세푸스 문제 2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] ar) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int index = 1;

        SegmentTree segmentTree = new SegmentTree(N);
        segmentTree.init(1, N, 1);

        sb.append("<");

        for (int i = 0; i < N; i++) {
            int size = N - i;
            index += (K - 1);

            if (index % size == 0) {
                index = size;
            } else if (index > size) {
                index %= size;
            }

            int delIndex = segmentTree.query(1, N, 1, index);

            segmentTree.update(1, N, 1, delIndex);

            if (i == N - 1) {
                sb.append(delIndex);
            } else {
                sb.append(delIndex).append(", ");
            }
        }

        sb.append(">");

        System.out.println(sb);

    }

}

class SegmentTree {

    long tree[];
    int treeSize;

    public SegmentTree(int arrSize) {
        int h = (int) Math.ceil(Math.log(arrSize) / Math.log(2));

        treeSize = (int) Math.pow(2, h + 1);
        tree = new long[treeSize];
    }

    public long init(int start, int end, int treeIndex) {
        if (start == end) {
            return tree[treeIndex] = 1;
        }

        return tree[treeIndex] = init(start, (start + end) / 2, treeIndex * 2)
                + init((start + end) / 2 + 1, end, treeIndex * 2 + 1);
    }

    public int update(int start, int end, int treeIndex, int delIndex) {
        tree[treeIndex]--;
        if (start == end) {
            return 0;
        } else {
            int mid = (start + end) / 2;
            if (delIndex <= mid) {
                return update(start, mid, treeIndex * 2, delIndex);
            } else {
                return update(mid + 1, end, treeIndex * 2 + 1, delIndex);
            }
        }
    }

    // 다음 순서에 해당하는 번호 받아오기
    public int query(int start, int end, int treeIndex, int order) {
        if (start == end) {
            return start;
        }
        int mid = (start + end) / 2;

        if (order <= tree[2 * treeIndex]) {
            return query(start, mid, treeIndex * 2, order);
        } else {
            return query(mid + 1, end, treeIndex * 2 + 1, (int) (order - tree[treeIndex * 2]));
        }
    }
}
